fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalCTEProducer.java

```diff
@@ -0,0 +1,143 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.trees.plans.logical;
+
+import org.apache.doris.nereids.memo.GroupExpression;
+import org.apache.doris.nereids.properties.LogicalProperties;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.PlanType;
+import org.apache.doris.nereids.trees.plans.visitor.PlanVisitor;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * LogicalCTEProducer
+ */
+public class LogicalCTEProducer<CHILD_TYPE extends Plan>
+        extends LogicalUnary<CHILD_TYPE> {
+
+    private final int cteId;
+
+    private final List<Slot> projects;
+
+    private final boolean rewritten;
+
+    public LogicalCTEProducer(CHILD_TYPE child, int cteId) {
+        super(PlanType.LOGICAL_CTE_PRODUCER, child);
+        this.cteId = cteId;
+        this.projects = Collections.emptyList();
+        this.rewritten = false;
+    }
+
+    public LogicalCTEProducer(Optional<GroupExpression> groupExpression,
+                              Optional<LogicalProperties> logicalProperties, CHILD_TYPE child, int cteId,
+                              List<Slot> projects, boolean rewritten) {
+        super(PlanType.LOGICAL_CTE_PRODUCER, groupExpression, logicalProperties, child);
+        this.cteId = cteId;
+        this.projects = ImmutableList.copyOf(java.util.Objects.requireNonNull(projects,
```

Comment:
import java.util.Objects
```suggestion
        this.projects = ImmutableList.copyOf(Objects.requireNonNull(projects,
```